// @tirke/node-cache-manager-ioredis/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/node-cache-manager-ioredis/src/index.ts
var src_exports = {};
__export(src_exports, {
  ioRedisStore: () => ioRedisStore
});
module.exports = __toCommonJS(src_exports);

// packages/node-cache-manager-ioredis/src/lib/node-cache-manager-ioredis.ts
var import_ioredis = __toESM(require("ioredis"));
var getVal = (value) => JSON.stringify(value) || '"undefined"';
var RedisStore = class {
  client;
  internalTtl;
  isCacheable;
  constructor(args) {
    this.internalTtl = args.ttl;
    this.isCacheable = args.isCacheable || ((value) => value !== void 0 && value !== null);
    if (args.instanceConfig) {
      this.client = new import_ioredis.default(args.instanceConfig);
    }
    if (args.clusterConfig) {
      this.client = new import_ioredis.default.Cluster(args.clusterConfig.nodes, args.clusterConfig.options);
    }
    if (args.redisInstance) {
      this.client = args.redisInstance;
    }
    if (!args.instanceConfig && !args.clusterConfig && !args.redisInstance) {
      this.client = new import_ioredis.default(args);
    }
  }
  async set(key, value, ttl) {
    if (!this.isCacheable(value)) {
      throw new Error(`${value} is not cacheable`);
    }
    const actualTtl = ttl ?? this.internalTtl;
    if (actualTtl) {
      await this.client.setex(key, actualTtl, getVal(value));
    } else {
      await this.client.set(key, getVal(value));
    }
  }
  async get(key) {
    const value = await this.client.get(key);
    if (value === null || value === void 0) {
      return void 0;
    }
    return JSON.parse(value);
  }
  async mset(args, ttl) {
    const actualTtl = ttl ?? this.internalTtl;
    if (actualTtl) {
      const multi = this.client.multi();
      for (const [key, value] of args) {
        if (!this.isCacheable(value)) {
          throw new Error(`${value} is not cacheable`);
        }
        multi.setex(key, actualTtl, getVal(value));
      }
      await multi.exec();
      return;
    }
    await this.client.mset(
      args.flatMap(([key, value]) => {
        if (!this.isCacheable(value)) {
          throw new Error(`${value} is not cacheable`);
        }
        return [key, getVal(value)];
      })
    );
  }
  async mget(...args) {
    const values = await this.client.mget(args);
    return values.map((val) => val === null || val === void 0 ? void 0 : JSON.parse(val));
  }
  async mdel(...args) {
    await this.client.del(args);
  }
  async del(key) {
    await this.client.del(key);
  }
  async reset() {
    await this.client.flushall();
  }
  async keys(pattern = "*") {
    return this.client.keys(pattern);
  }
  async ttl(key) {
    return this.client.ttl(key);
  }
};
var ioRedisStore = (args) => {
  if (!args) {
    return new RedisStore({});
  }
  return new RedisStore(args);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ioRedisStore
});
